lib/browser-specific.js

```diff
@@ -0,0 +1,247 @@
+  // tests with thousands of subtests from dominating the results.
+  if (browserTests.every(tests => !tests.subtests || tests.subtests.length == 0)) {
+    let failed = [];
+    for (let i = 0; i < browserTests.length; i++) {
+      // TODO: Determine exact criteria for a BSF at a test level.
+      if (browserTests[i].status !== 'PASS') {
+        failed.push(i);
+      }
+    }
+
+    if (failed.length == 1) {
+      scores[failed[0]] += 1;
+    }
+  } else if (browserTests.every(t => t.subtests && t.subtests.length > 0)) {
+    const comparator = (s1, s2) => (s1.name > s2.name) ? 1 : -1;
+    const browserSubtests = browserTests.map(
+        tests => new IteratorHelper(tests.subtests, comparator));
+
+    // To avoid errors from summing small floats, we do a full count of
+    // browser-specific subtest failures first, then divide by the number of
+    // subtests later to get the score (see the note on normalization above).
+    let denominator = 0;
+    let counts = new Array(browserSubtests.length).fill(0);
+
+    while (browserSubtests.every(subtests => subtests.hasNext())) {
+      // We only consider subtests that have reports for all browsers. A browser
+      // may be missing a result for a given subtest due to e.g. a timeout in an
+      // earlier subtest, and it would be wrong to penalize it for that.
+      const name = browserSubtests[0].currentValue().name;
+      if (browserSubtests.some(s => s.currentValue().name != name)) {
+        let smallestKey = name;
+        let smallestIdx = 0;
+        for (let i = 1; i < browserSubtests.length; i++) {
+          if (browserSubtests[i].currentValue().name < smallestKey) {
+            smallestKey = browserSubtests[i].currentValue().name;
+            smallestIdx = i;
+          }
+        }
+        browserSubtests[smallestIdx].next();
+        continue;
+      }
+
+      // The iterators are all aligned at the same subtest, so score it!
+      //
+      // NOTE: There actually (rarely) exist distinct subtests with the same
+      // name in the data, usually because of unprintable characters. This can
+      // influence the result, as we will squash such subtests and may even
+      // mismatch results (i.e. if some browser has results for one
+      // duplicate-named subtest but not another).
+      //
+      // Overall the impact is minor; it at most affects a fraction of a single
+      // test, so less than 1 point of the final score per affected test.
+      denominator += 1;
+
+      let failed = [];
+      for (let i = 0; i < browserSubtests.length; i++) {
+        // TODO: Determine exact criteria for a BSF at a subtest level.
+        if (browserSubtests[i].currentValue().status == 'FAIL') {
+          failed.push(i);
+        }
+      }
+      if (failed.length == 1) {
+        counts[failed[0]] += 1;
+      }
+
+      browserSubtests.forEach(s => s.next());
+    }
+
+    for (let i = 0; i < counts.length; i++) {
+      if (counts[i] > 0) {
```

Comment:
I'm not sure why we need to check this. It's fine to have 0 as the numerator, right?